adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1028 @@
+const DefaultUnknownNetworkType = 0
+const TimeFormat = "2006-01-02 15:04:05.000"
+const DefaultModelName = "HUAWEI"
+
+// creative type
+const (
+	text        int32 = 1
+	bigPicture  int32 = 2
+	bigPicture2 int32 = 3
+	gif         int32 = 4
+	// used for magazine lock screen
+	noSpecificCreativeType int32 = 5
+	videoText              int32 = 6
+	smallPicture           int32 = 7
+	threeSmallPicturesText int32 = 8
+	video                  int32 = 9
+	iconText               int32 = 10
+	// RewardedVideo only
+	videoWithPicturesText int32 = 11
+	windowAdvertisement   int32 = 13
+)
+
+// ads events
+const (
+	click         string = "click"
+	imp           string = "imp"
+	userClose     string = "userclose"
+	download      string = "download"
+	install       string = "install"
+	downloadstart string = "downloadstart"
+	playStart     string = "playStart"
+	playEnd       string = "playEnd"
+	playResume    string = "playResume"
+	playPause     string = "playPause"
+	appOpen       string = "appOpen"
+)
+
+// ads type
+const (
+	banner        int32 = 8
+	native        int32 = 3
+	rewardedVideo int32 = 7
+	splash        int32 = 1
+	interstitial  int32 = 12
+	roll          int32 = 60
+)
+
+type HuaweiAdsAdapter struct {
+	endpoint string
+}
+
+func (a *HuaweiAdsAdapter) MakeRequests(openRTBRequest *openrtb2.BidRequest,
+	reqInfo *adapters.ExtraRequestInfo) (requestsToBidder []*adapters.RequestData, errs []error) {
+	if openRTBRequest == nil {
+		errs = append(errs, &errortypes.BadInput{
+			Message: "MakeRequests: openRTBRequest is nil",
+		})
+		return nil, errs
+	}
+
+	numRequests := len(openRTBRequest.Imp)
+	if numRequests == 0 || openRTBRequest.Imp == nil {
+		errs = append(errs, &errortypes.BadInput{
+			Message: "MakeRequests: No impression in the bid request",
+		})
+		return nil, errs
+	}
+
+	var request HuaweiAdsRequest
+	var header http.Header
+	var multislot = make([]Adslot30, 0, numRequests)
+
+	var huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds
+	var err1 error
+	for index := 0; index < numRequests; index++ {
+		huaweiAdsImpExt, err1 = unmarshalExtImpHuaweiAds(&openRTBRequest.Imp[index])
+
+		if err1 != nil {
+			errs = append(errs, err1)
+			return nil, errs
+		}
+
+		adslot30, err := getHuaweiAdsReqAdslot30(huaweiAdsImpExt, &openRTBRequest.Imp[index], openRTBRequest)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+
+		multislot = append(multislot, adslot30)
+	}
+	request.Multislot = multislot
+
+	if err := getHuaweiAdsReqJson(&request, openRTBRequest, huaweiAdsImpExt); err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	//	our request header's Authorization is changing by time, cannot verify by a certain string,
+	//	use isAddAuthorization = false only when run testcase,
+	//	isAddAuthorization = true when we request for a new ad from Huawei Ads ADX.
+	var isAddAuthorization = true
```

Comment:
The function  getDigestAuthorization  here has been tested by our testers, and there is no problem with the access. Since the output value changes according to time, I have to exclude it from the unit test.